### PR TITLE
optimize idwt extrapolate decoder with neon

### PR DESCRIPTION
**Pull request by nero-ruijie**

### CODE REVIEW

**Comment by hardening** [comment]
@freerdp-bot can you give a test please ?

**Comment by mfleisz** [comment]
@nero-ruijie there are some formatting issues with your PR - please run clang-format on the changed files

**Comment by nero-ruijie** [comment]
> @nero-ruijie there are some formatting issues with your PR - please run clang-format on the changed files

i've tried clang-format which resulted in original codes also been impacted.
So, just format the additional code or commit the entire updates?

**Comment by hardening** [comment]
> > @nero-ruijie there are some formatting issues with your PR - please run clang-format on the changed files
> 
> i've tried clang-format which resulted in original codes also been impacted. So, just format the additional code or commit the entire updates?

Only the files you've changed.

**Comment by nero-ruijie** [comment]
> > > @nero-ruijie there are some formatting issues with your PR - please run clang-format on the changed files
> > 
> > 
> > i've tried clang-format which resulted in original codes also been impacted. So, just format the additional code or commit the entire updates?
> 
> Only the files you've changed.

done

**Comment by nero-ruijie** [comment]
> @nero-ruijie please pack your changes in 1 commit (no "format code" commit)

done

**Comment by hardening** [comment]
@freerdp-bot can you test please ?

File: libfreerdp/codec/progressive.c [hunk 1]
```diff
@@ -2810,6 +2815,9 @@ PROGRESSIVE_CONTEXT* progressive_context_new_ex(BOOL Compressor, UINT32 Threadin
 	progressive->rfx_context = rfx_context_new_ex(Compressor, ThreadingFlags);
 	if (!progressive->rfx_context)
 		goto fail;
+	if (!progressive->rfx_context->dwt_2d_extrapolate_decode)
```
Comment (akallabeth) [inline libfreerdp/codec/progressive.c:2818]: wrong place, move to where `dwt_2d_decode` is initialized to default (e.g. `rfx_context_new_ex`)

File: libfreerdp/codec/rfx_neon.c [hunk 2]
```diff
@@ -225,6 +225,292 @@ static void rfx_dwt_2d_decode_NEON(INT16* buffer, INT16* dwt_buffer)
 	rfx_dwt_2d_decode_block_NEON(buffer, dwt_buffer, 32);
 }
 
+static __inline void __attribute__((__gnu_inline__, __always_inline__, __artificial__))
```
Comment (akallabeth) [inline libfreerdp/codec/rfx_neon.c:228]: we have the `INLINE` macro for this in `FreeRDP`

File: libfreerdp/codec/rfx_neon.c [hunk 3]
```diff
@@ -225,6 +225,292 @@ static void rfx_dwt_2d_decode_NEON(INT16* buffer, INT16* dwt_buffer)
...
+rfx_idwt_extrapolate_horiz_neon(const INT16* pLowBand, size_t nLowStep, const INT16* pHighBand,
```
Comment (akallabeth) [inline libfreerdp/codec/rfx_neon.c:229]: you might want to experiment with the `restrict` keyword for optimization here

File: libfreerdp/codec/rfx_neon.c [hunk 4]
```diff
@@ -225,6 +225,292 @@ static void rfx_dwt_2d_decode_NEON(INT16* buffer, INT16* dwt_buffer)
...
+		{
+			*dst_ptr = *l_ptr;
+			l_ptr += 1;
+			dst_ptr += 1;
+		}
+	}
+}
+
+static __inline void __attribute__((__gnu_inline__, __always_inline__, __artificial__))
+rfx_idwt_extrapolate_vert_neon(const INT16* pLowBand, size_t nLowStep, const INT16* pHighBand,
```
Comment (akallabeth) [inline libfreerdp/codec/rfx_neon.c:313]: you might want to check the `restrict` keyword for arguments here

File: libfreerdp/codec/rfx_neon.c [hunk 5]
```diff
@@ -225,6 +225,292 @@ static void rfx_dwt_2d_decode_NEON(INT16* buffer, INT16* dwt_buffer)
...
+static INLINE size_t prfx_get_band_h_count(size_t level)
+{
+	if (level == 1)
+		return (64 >> 1) - 1;
+	else
+		return (64 + (1 << (level - 1))) >> level;
+}
+
+static __inline void __attribute__((__gnu_inline__, __always_inline__, __artificial__))
+rfx_dwt_2d_decode_extrapolate_block_neon(INT16* buffer, INT16* temp, size_t level)
```
Comment (akallabeth) [inline libfreerdp/codec/rfx_neon.c:469]: use macro `INLINE`

File: libfreerdp/codec/rfx_neon.c [hunk 6]
```diff
@@ -225,6 +225,292 @@ static void rfx_dwt_2d_decode_NEON(INT16* buffer, INT16* dwt_buffer)
...
+	/* horizontal (LH + HH -> H) */
+	rfx_idwt_extrapolate_horiz_neon(LH, nBandL, HH, nBandH, H, nDstStepX, nBandL, nBandH, nBandH);
+
+	/* vertical (L + H -> LL) */
+	rfx_idwt_extrapolate_vert_neon(L, nDstStepX, H, nDstStepX, LLx, nDstStepY, nBandL, nBandH,
+	                               nBandL + nBandH);
+}
+
+static void rfx_dwt_2d_extrapolate_decode_neon(INT16* buffer, INT16* temp)
+{
```
Comment (akallabeth) [inline libfreerdp/codec/rfx_neon.c:508]: `WINPR_ASSERT` arguments to better detect `API` misuse

**Review by akallabeth — CHANGES_REQUESTED** [review]

**Comment by akallabeth** [comment]
@nero-ruijie thank you for this contribution. I just did a small review (did not do thorough tests) and the changes look like a nice addition, but a few things need adjustment first.

**Comment by nero-ruijie** [comment]
@freerdp-bot test

**Comment by nero-ruijie** [comment]
> @nero-ruijie thank you for this contribution. I just did a small review (did not do thorough tests) and the changes look like a nice addition, but a few things need adjustment first.
@akallabeth I have updated the code as you commented. Pls check again.

File: libfreerdp/codec/progressive.c [hunk 7]
```diff
@@ -835,6 +835,13 @@ static INLINE void progressive_rfx_dwt_2d_decode_block(INT16* buffer, INT16* tem
 	                       nBandL + nBandH);
 }
 
+void rfx_dwt_2d_extrapolate_decode(INT16* buffer, INT16* temp)
+{
+	progressive_rfx_dwt_2d_decode_block(&buffer[3807], temp, 3);
```
Comment (akallabeth) [inline libfreerdp/codec/progressive.c:840]: `WINPR_ASSERT` arguments before dereference

File: libfreerdp/codec/progressive.c [hunk 8]
```diff
@@ -862,9 +869,7 @@ static INLINE int progressive_rfx_dwt_2d_decode(PROGRESSIVE_CONTEXT* progressive
 	}
 	else
 	{
-		progressive_rfx_dwt_2d_decode_block(&buffer[3807], temp, 3);
-		progressive_rfx_dwt_2d_decode_block(&buffer[3007], temp, 2);
-		progressive_rfx_dwt_2d_decode_block(&buffer[0], temp, 1);
+		progressive->rfx_context->dwt_2d_extrapolate_decode(buffer, temp);
```
Comment (akallabeth) [inline libfreerdp/codec/progressive.c:872]: `WINPR_ASSERT` function pointer

File: libfreerdp/codec/rfx_neon.c [hunk 9]
```diff
@@ -225,6 +225,296 @@ static void rfx_dwt_2d_decode_NEON(INT16* buffer, INT16* dwt_buffer)
 	rfx_dwt_2d_decode_block_NEON(buffer, dwt_buffer, 32);
 }
 
+static INLINE void __attribute__((__gnu_inline__, __always_inline__, __artificial__))
```
Comment (akallabeth) [inline libfreerdp/codec/rfx_neon.c:228]: remove `__attribute__`, we try not to assume any compiler (well, not 100% possible but this is already inline, so the keywords are mostly for debugging, right?)

File: libfreerdp/codec/rfx_neon.c [hunk 10]
```diff
@@ -225,6 +225,296 @@ static void rfx_dwt_2d_decode_NEON(INT16* buffer, INT16* dwt_buffer)
...
+rfx_idwt_extrapolate_horiz_neon(const INT16* restrict pLowBand, size_t nLowStep,
```
Comment (akallabeth) [inline libfreerdp/codec/rfx_neon.c:229]: `WINPR_ASSERT` function arguments (pointers)

File: libfreerdp/codec/rfx_neon.c [hunk 11]
```diff
@@ -225,6 +225,296 @@ static void rfx_dwt_2d_decode_NEON(INT16* buffer, INT16* dwt_buffer)
...
+		else
+		{
+			*dst_ptr = *l_ptr;
+			l_ptr += 1;
+			dst_ptr += 1;
+		}
+	}
+}
+
+static INLINE void __attribute__((__gnu_inline__, __always_inline__, __artificial__))
```
Comment (akallabeth) [inline libfreerdp/codec/rfx_neon.c:313]: remove the `__attribute__`

File: libfreerdp/codec/rfx_neon.c [hunk 12]
```diff
@@ -225,6 +225,296 @@ static void rfx_dwt_2d_decode_NEON(INT16* buffer, INT16* dwt_buffer)
...
+rfx_idwt_extrapolate_vert_neon(const INT16* restrict pLowBand, size_t nLowStep,
```
Comment (akallabeth) [inline libfreerdp/codec/rfx_neon.c:314]: `WINPR_ASSERT` arguments

File: libfreerdp/codec/rfx_neon.c [hunk 13]
```diff
@@ -225,6 +225,296 @@ static void rfx_dwt_2d_decode_NEON(INT16* buffer, INT16* dwt_buffer)
...
+
+static INLINE size_t prfx_get_band_h_count(size_t level)
+{
+	if (level == 1)
+		return (64 >> 1) - 1;
+	else
+		return (64 + (1 << (level - 1))) >> level;
+}
+
+static INLINE void __attribute__((__gnu_inline__, __always_inline__, __artificial__))
```
Comment (akallabeth) [inline libfreerdp/codec/rfx_neon.c:470]: remove `__attribute__`

File: libfreerdp/codec/rfx_neon.c [hunk 14]
```diff
@@ -225,6 +225,296 @@ static void rfx_dwt_2d_decode_NEON(INT16* buffer, INT16* dwt_buffer)
...
+                                INT16* restrict pDstBand, size_t nDstStep, size_t nLowCount,
+                                size_t nHighCount, size_t nDstCount)
+{
+	int y, n;
+	INT16* l_ptr = pLowBand;
+	INT16* h_ptr = pHighBand;
+	INT16* dst_ptr = pDstBand;
+	size_t batchSize = (nLowCount + nHighCount) >> 1;
+
+	for (y = 0; y < nDstCount; y++)
```
Comment (akallabeth) [inline libfreerdp/codec/rfx_neon.c:240]: we´re using modern `C` now, you can declare the variable in the loop like `for (size_t y=0; y<nDstCount; y++)`
also ensure the type matches (you have `int y` and `size_t nDstCount` here)

File: libfreerdp/codec/rfx_neon.c [hunk 15]
```diff
@@ -225,6 +225,296 @@ static void rfx_dwt_2d_decode_NEON(INT16* buffer, INT16* dwt_buffer)
...
+	{
+		/* Even coefficients */
+		for (n = 0; n < batchSize; n += 8)
```
Comment (akallabeth) [inline libfreerdp/codec/rfx_neon.c:243]: we´re using modern `C` now, you can declare the variable in the loop like `for (size n = 0; n < batchSize; n += 8)`
also ensure the type matches (you have `int y` and `size_t batchSize` here)

File: libfreerdp/codec/rfx_neon.c [hunk 16]
```diff
@@ -225,6 +225,296 @@ static void rfx_dwt_2d_decode_NEON(INT16* buffer, INT16* dwt_buffer)
...
+		if (n < 32)
+		{
+			*l_ptr -= *(h_ptr - 1);
+		}
+
+		l_ptr -= batchSize;
+		h_ptr -= batchSize;
+
+		/* Odd coefficients */
+		for (n = 0; n < batchSize; n += 8)
```
Comment (akallabeth) [inline libfreerdp/codec/rfx_neon.c:277]: we´re using modern `C` now, you can declare the variable in the loop like `for (size n = 0; n < batchSize; n += 8)`
also ensure the type matches (you have `int y` and `size_t batchSize` here)

File: libfreerdp/codec/rfx_neon.c [hunk 17]
```diff
@@ -225,6 +225,296 @@ static void rfx_dwt_2d_decode_NEON(INT16* buffer, INT16* dwt_buffer)
...
+{
+	int x, n;
+	INT16* l_ptr = pLowBand;
+	INT16* h_ptr = pHighBand;
+	INT16* dst_ptr = pDstBand;
+	int batchSize = (nDstCount >> 3) << 3;
+	int forceBandSize = (nLowCount + nHighCount) >> 1;
+
+	/* Even coefficients */
+	for (n = 0; n < forceBandSize; n++)
```
Comment (akallabeth) [inline libfreerdp/codec/rfx_neon.c:327]: we´re using modern `C` now, you can declare the variable in the loop like `for (size n = 0; n < forceBandSize; n ++)`
also ensure the type matches (you have `int y` and `size_t forceBandSize` here)

File: libfreerdp/codec/rfx_neon.c [hunk 18]
```diff
@@ -225,6 +225,296 @@ static void rfx_dwt_2d_decode_NEON(INT16* buffer, INT16* dwt_buffer)
...
+			h_ptr += 1;
+			dst_ptr += 1;
+		}
+	}
+
+	h_ptr = pHighBand;
+	dst_ptr = pDstBand + nDstStep;
+
+	/* Odd coefficients */
+	for (n = 0; n < forceBandSize; n++)
```
Comment (akallabeth) [inline libfreerdp/codec/rfx_neon.c:402]: we´re using modern `C` now, you can declare the variable in the loop like `for (size n = 0; n < forceBandSize; n ++)`
also ensure the type matches (you have `int y` and `size_t forceBandSize` here)

File: libfreerdp/codec/rfx_neon.c [hunk 19]
```diff
@@ -225,6 +225,296 @@ static void rfx_dwt_2d_decode_NEON(INT16* buffer, INT16* dwt_buffer)
...
+rfx_idwt_extrapolate_vert_neon(const INT16* restrict pLowBand, size_t nLowStep,
+                               const INT16* restrict pHighBand, size_t nHighStep,
+                               INT16* restrict pDstBand, size_t nDstStep, size_t nLowCount,
+                               size_t nHighCount, size_t nDstCount)
+{
+	int x, n;
+	INT16* l_ptr = pLowBand;
+	INT16* h_ptr = pHighBand;
+	INT16* dst_ptr = pDstBand;
+	int batchSize = (nDstCount >> 3) << 3;
```
Comment (akallabeth) [inline libfreerdp/codec/rfx_neon.c:323]: 1. wrong data type, truncation from `size_t` to `int`
2. can also be written `const size_t batchSize = nDstCount & ~0x07` or `const size_t batchSizxe = nDstCount & FFFFFFFFFFFFFFF8`

File: libfreerdp/codec/rfx_neon.c [hunk 20]
```diff
@@ -225,6 +225,296 @@ static void rfx_dwt_2d_decode_NEON(INT16* buffer, INT16* dwt_buffer)
...
+	int forceBandSize = (nLowCount + nHighCount) >> 1;
+
+	/* Even coefficients */
+	for (n = 0; n < forceBandSize; n++)
+	{
+		for (x = 0; x < batchSize; x += 8)
```
Comment (akallabeth) [inline libfreerdp/codec/rfx_neon.c:329]: we´re using modern C now, you can declare the variable in the loop like `for (size x = 0; x < batchSize; x++)`
also ensure the type matches (you have `int x` and `size_t` for `batchSize` here)

File: libfreerdp/codec/rfx_neon.c [hunk 21]
```diff
@@ -225,6 +225,296 @@ static void rfx_dwt_2d_decode_NEON(INT16* buffer, INT16* dwt_buffer)
...
+			h_ptr += 1;
+			dst_ptr += 1;
+		}
+
+		dst_ptr += nDstStep;
+	}
+
+	if (forceBandSize < 32)
+	{
+		for (x = 0; x < batchSize; x += 8)
```
Comment (akallabeth) [inline libfreerdp/codec/rfx_neon.c:378]: we´re using modern C now, you can declare the variable in the loop like `for (size x = 0; x < batchSize; x++)`
also ensure the type matches (you have `int x` and `size_t` for `batchSize` here)

File: libfreerdp/codec/rfx_neon.c [hunk 22]
```diff
@@ -225,6 +225,296 @@ static void rfx_dwt_2d_decode_NEON(INT16* buffer, INT16* dwt_buffer)
...
+		}
+	}
+
+	h_ptr = pHighBand;
+	dst_ptr = pDstBand + nDstStep;
+
+	/* Odd coefficients */
+	for (n = 0; n < forceBandSize; n++)
+	{
+		for (x = 0; x < batchSize; x += 8)
```
Comment (akallabeth) [inline libfreerdp/codec/rfx_neon.c:404]: we´re using modern C now, you can declare the variable in the loop like `for (size x = 0; x < batchSize; x++)`
also ensure the type matches (you have `int x` and `size_t` for `batchSize` here)

File: libfreerdp/codec/rfx_neon.c [hunk 23]
```diff
@@ -225,6 +225,296 @@ static void rfx_dwt_2d_decode_NEON(INT16* buffer, INT16* dwt_buffer)
...
+static INLINE size_t prfx_get_band_h_count(size_t level)
+{
+	if (level == 1)
+		return (64 >> 1) - 1;
+	else
+		return (64 + (1 << (level - 1))) >> level;
+}
+
+static INLINE void __attribute__((__gnu_inline__, __always_inline__, __artificial__))
+rfx_dwt_2d_decode_extrapolate_block_neon(INT16* buffer, INT16* temp, size_t level)
```
Comment (akallabeth) [inline libfreerdp/codec/rfx_neon.c:471]: `WINPR_ASSERT` pointer arguments

File: libfreerdp/codec/rfx_neon.c [hunk 24]
```diff
@@ -225,6 +225,296 @@ static void rfx_dwt_2d_decode_NEON(INT16* buffer, INT16* dwt_buffer)
...
+	size_t nDstStepY;
+	INT16 *HL, *LH;
+	INT16 *HH, *LL;
+	INT16 *L, *H, *LLx;
+
+	const size_t nBandL = prfx_get_band_l_count(level);
+	const size_t nBandH = prfx_get_band_h_count(level);
+	size_t offset = 0;
+
+	HL = &buffer[offset];
```
Comment (akallabeth) [inline libfreerdp/codec/rfx_neon.c:483]: you can declare the variables mid code now (`C99`) and better make them `const` if you do not intend to modify them

File: libfreerdp/codec/rfx_neon.c [hunk 25]
```diff
@@ -225,6 +225,296 @@ static void rfx_dwt_2d_decode_NEON(INT16* buffer, INT16* dwt_buffer)
...
+			h_ptr += 8;
+			dst_ptr += 8;
+		}
+
+		if (nDstCount > batchSize)
+		{
+			int16_t h_n = (n == 31) ? *(h_ptr - nHighStep) : *h_ptr;
+			int16_t tmp_n = h_n + 1;
+			if (n == 0)
+			{
```
Comment (akallabeth) [inline libfreerdp/codec/rfx_neon.c:359]: no braces required for one liners

File: libfreerdp/codec/rfx_neon.c [hunk 26]
```diff
@@ -225,6 +225,296 @@ static void rfx_dwt_2d_decode_NEON(INT16* buffer, INT16* dwt_buffer)
...
+				tmp_n += h_n;
+			}
+			else if (n < 31)
+			{
```
Comment (akallabeth) [inline libfreerdp/codec/rfx_neon.c:363]: no braces required for one liners

File: libfreerdp/codec/rfx_neon.c [hunk 27]
```diff
@@ -225,6 +225,296 @@ static void rfx_dwt_2d_decode_NEON(INT16* buffer, INT16* dwt_buffer)
...
+			int16x8_t l_n = vld1q_s16(l_ptr);
+			int16x8_t h_n = vld1q_s16((n == 31) ? (h_ptr - nHighStep) : h_ptr);
+			int16x8_t tmp_n = vaddq_s16(h_n, vdupq_n_s16(1));
+
+			if (n == 0)
+			{
+				tmp_n = vaddq_s16(tmp_n, h_n);
+			}
+			else if (n < 31)
+			{
```
Comment (akallabeth) [inline libfreerdp/codec/rfx_neon.c:341]: no braces required for one liners

File: libfreerdp/codec/rfx_neon.c [hunk 28]
```diff
@@ -225,6 +225,296 @@ static void rfx_dwt_2d_decode_NEON(INT16* buffer, INT16* dwt_buffer)
...
+	{
+		for (x = 0; x < batchSize; x += 8)
+		{
+			// dst[2n] = l[n] - ((h[n-1] + h[n] + 1) >> 1);
+			int16x8_t l_n = vld1q_s16(l_ptr);
+			int16x8_t h_n = vld1q_s16((n == 31) ? (h_ptr - nHighStep) : h_ptr);
+			int16x8_t tmp_n = vaddq_s16(h_n, vdupq_n_s16(1));
+
+			if (n == 0)
+			{
```
Comment (akallabeth) [inline libfreerdp/codec/rfx_neon.c:337]: no braces required for one liners

File: libfreerdp/codec/rfx_neon.c [hunk 29]
```diff
@@ -225,6 +225,296 @@ static void rfx_dwt_2d_decode_NEON(INT16* buffer, INT16* dwt_buffer)
...
+		{
+			// dst[2n + 1] = (h[n] << 1) + ((dst[2n] + dst[2n + 2]) >> 1);
+			int16x8_t h_n = vld1q_s16(h_ptr);
+			h_n = vshlq_n_s16(h_n, 1);
+			int16x8x2_t dst_n;
+			dst_n.val[0] = vld1q_s16(l_ptr);
+			int16x8_t dst_n_p = vld1q_s16(l_ptr + 1);
+
+			if (n == 24)
+			{
```
Comment (akallabeth) [inline libfreerdp/codec/rfx_neon.c:287]: no braces required for one liners

File: libfreerdp/codec/rfx_neon.c [hunk 30]
```diff
@@ -225,6 +225,296 @@ static void rfx_dwt_2d_decode_NEON(INT16* buffer, INT16* dwt_buffer)
...
+			int16x8_t tmp_n = vaddq_s16(h_n, h_n_m);
+			tmp_n = vaddq_s16(tmp_n, vdupq_n_s16(1));
+			tmp_n = vshrq_n_s16(tmp_n, 1);
+			int16x8_t dst_n = vsubq_s16(l_n, tmp_n);
+			vst1q_s16(l_ptr, dst_n);
+			l_ptr += 8;
+			h_ptr += 8;
+		}
+		if (n < 32)
+		{
```
Comment (akallabeth) [inline libfreerdp/codec/rfx_neon.c:269]: no braces required for one liners

File: libfreerdp/codec/rfx_neon.c [hunk 31]
```diff
@@ -225,6 +225,296 @@ static void rfx_dwt_2d_decode_NEON(INT16* buffer, INT16* dwt_buffer)
...
+			int16x8_t h_n = vld1q_s16(h_ptr);
+			int16x8_t h_n_m = vld1q_s16(h_ptr - 1);
+
+			if (n == 0)
+			{
+				int16_t first = vgetq_lane_s16(h_n_m, 1);
+				h_n_m = vsetq_lane_s16(first, h_n_m, 0);
+			}
+			else if (n == 24)
+			{
```
Comment (akallabeth) [inline libfreerdp/codec/rfx_neon.c:256]: no braces required for one liners

File: libfreerdp/codec/rfx_neon.c [hunk 32]
```diff
@@ -225,6 +225,296 @@ static void rfx_dwt_2d_decode_NEON(INT16* buffer, INT16* dwt_buffer)
...
+		/* Even coefficients */
+		for (n = 0; n < batchSize; n += 8)
+		{
+			// dst[2n] = l[n] - ((h[n-1] + h[n] + 1) >> 1);
+			int16x8_t l_n = vld1q_s16(l_ptr);
+			int16x8_t h_n = vld1q_s16(h_ptr);
+			int16x8_t h_n_m = vld1q_s16(h_ptr - 1);
+
+			if (n == 0)
+			{
```
Comment (akallabeth) [inline libfreerdp/codec/rfx_neon.c:251]: no braces required for one liners

**Review by akallabeth — CHANGES_REQUESTED** [review]
some more in depth review.

code looks good (except some assertions missing and loop variable type mismatches).
will check functionality later when I have time, raspberry is slow on compile :)

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by nero-ruijie** [comment]
> some more in depth review.
> 
> code looks good (except some assertions missing and loop variable type mismatches). will check functionality later when I have time, raspberry is slow on compile :)

Code updated again as you commented with one exception:
H/L/HL/... band buffer ptr is keept as non-const due to both buffer(idwt data src and result buffer) and temp (buffer) will be updated in the final, especially in rfx_idwt_extrapolate_horiz_neon, medial data is stored in pLowBand for performance reason(also refer to rfx_dwt_2d_decode_block_horiz_NEON) where pLowBand is the input data buffer and is marked "const" in progressive_rfx_idwt_x.

**Review by akallabeth — APPROVED** [review]
LGTM

**Comment by akallabeth** [comment]
@freerdp-bot test

**Pull request merged**